File: binding/include/libqscroller.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qscroller.html
pub const qscroller = struct {

    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QScroller_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qscroller.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn HasScroller(target: ?*anyopaque) bool {
        return qtc.QScroller_HasScroller(@ptrCast(target));
    }

    pub fn Scroller(target: ?*anyopaque) QtC.QScroller {
        return qtc.QScroller_Scroller(@ptrCast(target));
    }

    pub fn Scroller2(target: ?*anyopaque) QtC.QScroller {
        return qtc.QScroller_Scroller2(@ptrCast(target));
    }

    pub fn GrabGesture(target: ?*anyopaque) i64 {
        return qtc.QScroller_GrabGesture(@ptrCast(target));
    }

    pub fn GrabbedGesture(target: ?*anyopaque) i64 {
        return qtc.QScroller_GrabbedGesture(@ptrCast(target));
    }

    pub fn UngrabGesture(target: ?*anyopaque) void {
        qtc.QScroller_UngrabGesture(@ptrCast(target));
    }

    pub fn ActiveScrollers() []const u8 {
        return qtc.QScroller_ActiveScrollers();
    }

    pub fn Target(self: ?*anyopaque, ) QtC.QObject {
        return qtc.QScroller_Target(@ptrCast(self));
    }

    pub fn State(self: ?*anyopaque, ) i32 {
        return qtc.QScroller_State(@ptrCast(self));
    }

    pub fn HandleInput(self: ?*anyopaque, input: i32, position: ?*anyopaque) bool {
        return qtc.QScroller_HandleInput(@ptrCast(self), @intCast(input), @ptrCast(position));
    }

    pub fn Stop(self: ?*anyopaque, ) void {
        qtc.QScroller_Stop(@ptrCast(self));
    }

    pub fn Velocity(self: ?*anyopaque, ) QtC.QPointF {
        return qtc.QScroller_Velocity(@ptrCast(self));
    }

    pub fn FinalPosition(self: ?*anyopaque, ) QtC.QPointF {
        return qtc.QScroller_FinalPosition(@ptrCast(self));
    }

    pub fn PixelPerMeter(self: ?*anyopaque, ) QtC.QPointF {
        return qtc.QScroller_PixelPerMeter(@ptrCast(self));
    }

    pub fn ScrollerProperties(self: ?*anyopaque, ) QtC.QScrollerProperties {
        return qtc.QScroller_ScrollerProperties(@ptrCast(self));
    }

    pub fn SetSnapPositionsX(self: ?*anyopaque, positions: []const u8) void {
        qtc.QScroller_SetSnapPositionsX(@ptrCast(self), @ptrCast(positions));
    }

    pub fn SetSnapPositionsX2(self: ?*anyopaque, first: f64, interval: f64) void {
        qtc.QScroller_SetSnapPositionsX2(@ptrCast(self), first, interval);
    }

    pub fn SetSnapPositionsY(self: ?*anyopaque, positions: []const u8) void {
        qtc.QScroller_SetSnapPositionsY(@ptrCast(self), @ptrCast(positions));
    }

    pub fn SetSnapPositionsY2(self: ?*anyopaque, first: f64, interval: f64) void {
        qtc.QScroller_SetSnapPositionsY2(@ptrCast(self), first, interval);
    }

    pub fn SetScrollerProperties(self: ?*anyopaque, prop: ?*anyopaque) void {
        qtc.QScroller_SetScrollerProperties(@ptrCast(self), @ptrCast(prop));
    }

    pub fn ScrollTo(self: ?*anyopaque, pos: ?*anyopaque) void {
        qtc.QScroller_ScrollTo(@ptrCast(self), @ptrCast(pos));
    }

    pub fn ScrollTo2(self: ?*anyopaque, pos: ?*anyopaque, scrollTime: i32) void {
        qtc.QScroller_ScrollTo2(@ptrCast(self), @ptrCast(pos), scrollTime);
    }

    pub fn EnsureVisible(self: ?*anyopaque, rect: ?*anyopaque, xmargin: f64, ymargin: f64) void {
        qtc.QScroller_EnsureVisible(@ptrCast(self), @ptrCast(rect), xmargin, ymargin);
    }

    pub fn EnsureVisible2(self: ?*anyopaque, rect: ?*anyopaque, xmargin: f64, ymargin: f64, scrollTime: i32) void {
        qtc.QScroller_EnsureVisible2(@ptrCast(self), @ptrCast(rect), xmargin, ymargin, scrollTime);
    }

    pub fn ResendPrepareEvent(self: ?*anyopaque, ) void {
        qtc.QScroller_ResendPrepareEvent(@ptrCast(self));
    }

    pub fn StateChanged(self: ?*anyopaque, newstate: i32) void {
        qtc.QScroller_StateChanged(@ptrCast(self), @intCast(newstate));
    }

    pub fn OnStateChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QScroller_Connect_StateChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn ScrollerPropertiesChanged(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QScroller_ScrollerPropertiesChanged(@ptrCast(self), @ptrCast(param1));
    }

    pub fn OnScrollerPropertiesChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QScroller_Connect_ScrollerPropertiesChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QScroller_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qscroller.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QScroller_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qscroller.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn GrabGesture2(target: ?*anyopaque, gestureType: i32) i64 {
        return qtc.QScroller_GrabGesture2(@ptrCast(target), @intCast(gestureType));
    }

    pub fn HandleInput3(self: ?*anyopaque, input: i32, position: ?*anyopaque, timestamp: i64) bool {
        return qtc.QScroller_HandleInput3(@ptrCast(self), @intCast(input), @ptrCast(position), timestamp);
    }
};

/// https://doc.qt.io/qt-6/scroller.html#types
pub const enums = struct {
    pub const State = enum {
        pub const Inactive: i32 = 0;
        pub const Pressed: i32 = 1;
        pub const Dragging: i32 = 2;
        pub const Scrolling: i32 = 3;
    };

    pub const ScrollerGestureType = enum {
        pub const TouchGesture: i32 = 0;
        pub const LeftMouseButtonGesture: i32 = 1;
        pub const RightMouseButtonGesture: i32 = 2;
        pub const MiddleMouseButtonGesture: i32 = 3;
    };

    pub const Input = enum {
        pub const InputPress: i32 = 1;
        pub const InputMove: i32 = 2;
        pub const InputRelease: i32 = 3;
    };

};

File: binding/src/libqscroller.cpp
#include <QObject>
#include <QPointF>
#include <QRectF>
#include <QScroller>
#include <QScrollerProperties>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qscroller.h>
#include "libqscroller.h"
#include "libqscroller.hxx"

libqt_string QScroller_Tr(const char* s) {
	QString _ret = QScroller::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QScroller_HasScroller(QObject* target) {
	return QScroller::hasScroller(target);
}

QScroller* QScroller_Scroller(QObject* target) {
	return QScroller::scroller(target);
}

const QScroller* QScroller_Scroller2(const QObject* target) {
	return QScroller::scroller(target);
}

int QScroller_GrabGesture(QObject* target) {
	return QScroller::grabGesture(target);
}

int QScroller_GrabbedGesture(QObject* target) {
	return QScroller::grabbedGesture(target);
}

void QScroller_UngrabGesture(QObject* target) {
	QScroller::ungrabGesture(target);
}

libqt_list QScroller_ActiveScrollers() {
	return QScroller::activeScrollers();
}

QObject* QScroller_Target(const QScroller* self) {
	return self->target();
}

int QScroller_State(const QScroller* self) {
	return self->state();
}

bool QScroller_HandleInput(QScroller* self, int input, const QPointF* position) {
	return self->handleInput(static_cast<QScroller::Input>(input), *position);
}

void QScroller_Stop(QScroller* self) {
	self->stop();
}

QPointF* QScroller_Velocity(const QScroller* self) {
	return new QPointF(self->velocity());
}

QPointF* QScroller_FinalPosition(const QScroller* self) {
	return new QPointF(self->finalPosition());
}

QPointF* QScroller_PixelPerMeter(const QScroller* self) {
	return new QPointF(self->pixelPerMeter());
}

QScrollerProperties* QScroller_ScrollerProperties(const QScroller* self) {
	return new QScrollerProperties(self->scrollerProperties());
}

void QScroller_SetSnapPositionsX(QScroller* self, const libqt_list positions) {
	self->setSnapPositionsX(*positions);
}

void QScroller_SetSnapPositionsX2(QScroller* self, double first, double interval) {
	self->setSnapPositionsX(first, interval);
}

void QScroller_SetSnapPositionsY(QScroller* self, const libqt_list positions) {
	self->setSnapPositionsY(*positions);
}

void QScroller_SetSnapPositionsY2(QScroller* self, double first, double interval) {
	self->setSnapPositionsY(first, interval);
}

void QScroller_SetScrollerProperties(QScroller* self, const QScrollerProperties* prop) {
	self->setScrollerProperties(*prop);
}

void QScroller_ScrollTo(QScroller* self, const QPointF* pos) {
	self->scrollTo(*pos);
}

void QScroller_ScrollTo2(QScroller* self, const QPointF* pos, int scrollTime) {
	self->scrollTo(*pos, scrollTime);
}

void QScroller_EnsureVisible(QScroller* self, const QRectF* rect, double xmargin, double ymargin) {
	self->ensureVisible(*rect, xmargin, ymargin);
}

void QScroller_EnsureVisible2(QScroller* self, const QRectF* rect, double xmargin, double ymargin, int scrollTime) {
	self->ensureVisible(*rect, xmargin, ymargin, scrollTime);
}

void QScroller_ResendPrepareEvent(QScroller* self) {
	self->resendPrepareEvent();
}

void QScroller_StateChanged(QScroller* self, int newstate) {
	self->stateChanged(static_cast<QScroller::State>(newstate));
}

void QScroller_Connect_StateChanged(QScroller* self, intptr_t slot) {
    void (*slotFunc)(QScroller*, int) = reinterpret_cast<void (*)(QScroller*, int)>(slot);
    QScroller::connect(self, &QScroller::stateChanged, [self, slotFunc](QScroller::State newstate) {
	slotFunc(self, newstate);
    });
}

void QScroller_ScrollerPropertiesChanged(QScroller* self, const QScrollerProperties* param1) {
	self->scrollerPropertiesChanged(*param1);
}

void QScroller_Connect_ScrollerPropertiesChanged(QScroller* self, intptr_t slot) {
    void (*slotFunc)(QScroller*, const QScrollerProperties*) = reinterpret_cast<void (*)(QScroller*, const QScrollerProperties*)>(slot);
    QScroller::connect(self, &QScroller::scrollerPropertiesChanged, [self, slotFunc](const QScrollerProperties& param1) {
	slotFunc(self, param1);
    });
}

libqt_string QScroller_Tr2(const char* s, const char* c) {
	QString _ret = QScroller::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QScroller_Tr3(const char* s, const char* c, int n) {
	QString _ret = QScroller::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QScroller_GrabGesture2(QObject* target, int gestureType) {
	return QScroller::grabGesture(target, static_cast<QScroller::ScrollerGestureType>(gestureType));
}

bool QScroller_HandleInput3(QScroller* self, int input, const QPointF* position, long long timestamp) {
	return self->handleInput(static_cast<QScroller::Input>(input), *position, timestamp);
}

